File: internal/tdd/tdd_test.go
package tdd_test

import (
	"fmt"
	"os"
	"path/filepath"
	"testing"

	"github.com/delightfulhammers/telesis/internal/tdd"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func setupTDDDir(t *testing.T) string {
	t.Helper()
	dir := t.TempDir()
	require.NoError(t, os.MkdirAll(filepath.Join(dir, "docs", "tdd"), 0o755))
	return dir
}

func TestCreateFirstTDD(t *testing.T) {
	rootDir := setupTDDDir(t)

	path, err := tdd.Create(rootDir, "config-loader")
	require.NoError(t, err)

	assert.Equal(t, filepath.Join(rootDir, "docs", "tdd", "TDD-001-config-loader.md"), path)
	assert.FileExists(t, path)

	content, err := os.ReadFile(path)
	require.NoError(t, err)
	assert.Contains(t, string(content), "# TDD-001: config-loader")
	assert.Contains(t, string(content), "## Overview")
	assert.Contains(t, string(content), "## Components")
	assert.Contains(t, string(content), "## Interfaces")
	assert.Contains(t, string(content), "## Data Model")
	assert.Contains(t, string(content), "## Open Questions")
}

func TestCreateSequentialTDDs(t *testing.T) {
	rootDir := setupTDDDir(t)

	path1, err := tdd.Create(rootDir, "first")
	require.NoError(t, err)
	assert.Contains(t, path1, "TDD-001-first.md")

	path2, err := tdd.Create(rootDir, "second")
	require.NoError(t, err)
	assert.Contains(t, path2, "TDD-002-second.md")
}

func TestCreateTDDWithExistingGap(t *testing.T) {
	rootDir := setupTDDDir(t)
	tddDir := filepath.Join(rootDir, "docs", "tdd")

	require.NoError(t, os.WriteFile(
		filepath.Join(tddDir, "TDD-010-existing.md"),
		[]byte("# TDD-010: existing\n"),
		0o644,
	))

	path, err := tdd.Create(rootDir, "next")
	require.NoError(t, err)
	assert.Contains(t, path, "TDD-011-next.md")
}

func TestCreateTDDRejectsEmptySlug(t *testing.T) {
	rootDir := setupTDDDir(t)

	_, err := tdd.Create(rootDir, "")
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "slug")
}

func TestCreateTDDRejectsInvalidSlug(t *testing.T) {
	tests := []struct {
		name string
		slug string
	}{
		{"spaces", "has spaces"},
		{"uppercase", "HasUpperCase"},
		{"special chars", "has/slashes"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			rootDir := setupTDDDir(t)
			_, err := tdd.Create(rootDir, tt.slug)
			assert.Error(t, err)
			assert.Contains(t, err.Error(), "slug")
		})
	}
}

func TestNextNumberEmptyDir(t *testing.T) {
	rootDir := setupTDDDir(t)

	num, err := tdd.NextNumber(filepath.Join(rootDir, "docs", "tdd"))
	require.NoError(t, err)
	assert.Equal(t, 1, num)
}

func TestNextNumberWithExisting(t *testing.T) {
	rootDir := setupTDDDir(t)
	tddDir := filepath.Join(rootDir, "docs", "tdd")

	for i := 1; i <= 3; i++ {
		require.NoError(t, os.WriteFile(
			filepath.Join(tddDir, fmt.Sprintf("TDD-%03d-test.md", i)),
			[]byte(fmt.Sprintf("# TDD-%03d: test\n", i)),
			0o644,
		))
	}

	num, err := tdd.NextNumber(tddDir)
	require.NoError(t, err)
	assert.Equal(t, 4, num)
}

func TestNextNumberIgnoresNonTDDFiles(t *testing.T) {
	rootDir := setupTDDDir(t)
	tddDir := filepath.Join(rootDir, "docs", "tdd")

	require.NoError(t, os.WriteFile(
		filepath.Join(tddDir, "README.md"),
		[]byte("# TDDs\n"),
		0o644,
	))

	num, err := tdd.NextNumber(tddDir)
	require.NoError(t, err)
	assert.Equal(t, 1, num)
}

File: src/tdd/tdd.test.ts
import { describe, it, expect } from "vitest";
import { mkdtempSync, mkdirSync, writeFileSync, readFileSync, existsSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { create, nextNumber } from "./tdd.js";

const makeTempDir = (): string =>
  mkdtempSync(join(tmpdir(), "telesis-tdd-test-"));

const setupTDDDir = (): string => {
  const dir = makeTempDir();
  mkdirSync(join(dir, "docs", "tdd"), { recursive: true });
  return dir;
};

describe("tdd", () => {
  it("creates first TDD", () => {
    const rootDir = setupTDDDir();
    const path = create(rootDir, "config-loader");

    expect(path).toBe(
      join(rootDir, "docs", "tdd", "TDD-001-config-loader.md"),
    );
    expect(existsSync(path)).toBe(true);

    const content = readFileSync(path, "utf-8");
    expect(content).toContain("# TDD-001: config-loader");
    expect(content).toContain("## Overview");
    expect(content).toContain("## Components");
    expect(content).toContain("## Interfaces");
    expect(content).toContain("## Data Model");
    expect(content).toContain("## Open Questions");
  });

  it("creates sequential TDDs", () => {
    const rootDir = setupTDDDir();

    const path1 = create(rootDir, "first");
    expect(path1).toContain("TDD-001-first.md");

    const path2 = create(rootDir, "second");
    expect(path2).toContain("TDD-002-second.md");
  });

  it("creates TDD with existing gap", () => {
    const rootDir = setupTDDDir();
    const tddDir = join(rootDir, "docs", "tdd");

    writeFileSync(
      join(tddDir, "TDD-010-existing.md"),
      "# TDD-010: existing\n",
    );

    const path = create(rootDir, "next");
    expect(path).toContain("TDD-011-next.md");
  });

  it("rejects empty slug", () => {
    const rootDir = setupTDDDir();
    expect(() => create(rootDir, "")).toThrow("slug");
  });

  it("rejects invalid slugs", () => {
    const invalidSlugs = [
      { name: "spaces", slug: "has spaces" },
      { name: "uppercase", slug: "HasUpperCase" },
      { name: "special chars", slug: "has/slashes" },
    ];

    for (const { slug } of invalidSlugs) {
      const rootDir = setupTDDDir();
      expect(() => create(rootDir, slug)).toThrow("slug");
    }
  });

  describe("nextNumber", () => {
    it("returns 1 for empty directory", () => {
      const rootDir = setupTDDDir();
      const num = nextNumber(join(rootDir, "docs", "tdd"));
      expect(num).toBe(1);
    });

    it("returns next with existing files", () => {
      const rootDir = setupTDDDir();
      const tddDir = join(rootDir, "docs", "tdd");

      for (let i = 1; i <= 3; i++) {
        writeFileSync(
          join(tddDir, `TDD-${String(i).padStart(3, "0")}-test.md`),
          `# TDD-${String(i).padStart(3, "0")}: test\n`,
        );
      }

      const num = nextNumber(tddDir);
      expect(num).toBe(4);
    });

    it("ignores non-TDD files", () => {
      const rootDir = setupTDDDir();
      const tddDir = join(rootDir, "docs", "tdd");

      writeFileSync(join(tddDir, "README.md"), "# TDDs\n");

      const num = nextNumber(tddDir);
      expect(num).toBe(1);
    });
  });
});
